File: shift_cipher.py
import substitution


def get_shift_map(alphabet: str, shift: int) -> dict[str, str]:
    map = { char: alphabet[(i + shift) % len(alphabet)] for i, char in enumerate(alphabet)}

    return map


def get_keyed_map(alphabet: str, key: str) -> dict[str, str]:
    map = {}
    i = 0
    for char in key:
        map[alphabet[i]] = char
        i += 1

    for char in alphabet:
        if i >= len(alphabet):
            break

        if char not in key:
            map[alphabet[i]] = char
            i += 1

    return map


def shift_msg_range(text: str, shift_range: tuple[int, int], alphabet: str) -> list[str]:
    shift_maps = [get_shift_map(alphabet, i) for i in range(shift_range[0], shift_range[1]) ]
    shifted_texts = substitution.get_substitution_over(text, shift_maps)

    return shifted_texts


def enumerate_shift(text: str, alphabet: str) -> list[str]:
    shift_range = (0, len(alphabet))
    texts: list[str] = shift_msg_range(text, shift_range, alphabet)
    return texts

File: crypt/app/util/shiftCipher.ts
import { getSubstitutionOver } from "./substitution";

function getShiftMap(alphabet: string, shift: number): Map<string, string> {
	const map = new Map;

	for (let i = 0; i < alphabet.length; i++) {

		const char = alphabet[i];
		const val = alphabet[(i + shift) % alphabet.length]

		map.set(char, val);
	}

	return map;
}

function getKeyedMap(alphabet: string, key: string): Map<string, string> {
	const map = new Map;

	let j = 0;
	for (let i = 0; i < key.length; i++) { 
		map.set(alphabet[j], key[i])
		j++;
	}

	for (let i = 0; i < alphabet.length; i++) {
		if (i >= alphabet.length) {
			break;
		}

		const char = alphabet[i];
		if (!key.includes(char)) {
			map.set(char, alphabet[j])
		}
	}

	return map;
}

export function shiftMsgRange(text: string, shiftRange: [number, number], alphabet: string): Array<string> {
	const shiftMaps: Array<Map<string,string>> = [];
	for (let i = shiftRange[0]; i < shiftRange[1]; i++) {
		shiftMaps.push(getShiftMap(alphabet, i));
	}

	const shiftedTexts = getSubstitutionOver(text, shiftMaps);

	return shiftedTexts;
}

export function enumerateShift(text: string, alphabet: string): Array<string> {
	const shiftRange: [number, number] = [0, alphabet.length];

	const texts = shiftMsgRange(text, shiftRange, alphabet);

	return texts;
}
